src/cancelq/posix/cancellable.c

```diff
@@ -0,0 +1,399 @@
+#include <stdlib.h>
+#include <string.h>
+#include <assert.h>
+#include <stdio.h>
+#include <inttypes.h>
+#include <pthread.h>
+
+#include "cancellable.h"
+
+#ifndef HEAP_CHECK
+#define HEAP_CHECK true // Set to false to save oodles of execution time.
+#endif
+
+#define BASE_HEAP_SIZE 25
+#define INITIAL_ID_COUNT 16
+
+#define UNUSED __attribute__((unused))
+
+// Normal events that can't be cancelled
+typedef struct
+{
+    const void *event;
+} event_t;
+
+// Events that can be cancelled, which includes events whose sending
+// can be delayed
+typedef struct
+{
+    event_t event;
+    cancellable_id_t id;
+    abs_time_t delivery_time;
+} cancellable_t;
+
+typedef cancellable_t heap_data_t;
+#include "heap.c" // Note the .c
+
+typedef enum
+{
+    ID_UNUSED = 0,
+    ID_DELIVERED,
+    ID_WAITING
+} id_state_t;
+
+struct event_queue_s
+{
+    heap_t heap;
+    // Current size of the array of IDs
+    cancellable_id_t idsize;
+
+    // Array of IDs for cancellation. The ID is an index into this array.
+    id_state_t *ids;
+
+    pthread_mutex_t lock;
+    pthread_cond_t empty;
+};
+
+UNUSED static void print_queue(event_queue_t *q)
+{
+    print_heap(&q->heap);
+}
+
+static bool empty_LH(const event_queue_t *q)
+{
+    return heap_empty(&q->heap);
+}
+
+static cancellable_id_t new_id(event_queue_t *q)
+{
+    id_state_t *new_ids;
+    cancellable_id_t new_size;
+    cancellable_id_t id;
+
+    for (id = 0; id < q->idsize; id++)
+    {
+        if (q->ids[id] == ID_UNUSED)
+        {
+            q->ids[id] = ID_WAITING;
+            return id;
+        }
+    }
+
+    // We're out of IDs. Double the size of the array for now.
+    new_size = q->idsize * 2;
+    new_ids = realloc(q->ids, sizeof(id_state_t) * new_size);
+    if (new_ids == NULL)
+        return NOT_CANCELLABLE;
+    q->ids = new_ids;
+
+    // Depends on the fact that we're doubling.
+    memset(&q->ids[id], ID_UNUSED, sizeof(q->ids[id]) * q->idsize);
+
+    q->idsize = new_size;
+
+    q->ids[id] = ID_WAITING;
+    return id;
+}
+
+static bool check_queue(const event_queue_t *q)
+{
+    return check_heap(&q->heap);
+}
+
+event_queue_t *eq_new(void)
+{
+    event_queue_t *q;
+
+    q = malloc(sizeof(*q));
+    if (q == NULL)
+        return q;
+    newheap(&q->heap, BASE_HEAP_SIZE);
+    q->idsize = INITIAL_ID_COUNT;
+
+    if (!check_heap(&q->heap))
```

Comment:
It catches failed allocation. Granted, it's not the cleanest way to test if `q->heap == NULL`, but I figured it might be possible for other kinds of things to go wrong in the future and it could catch those too.